pkg/system/filesys_windows.go

```diff
@@ -164,15 +165,55 @@ func OpenFileSequential(name string, flag int, _ os.FileMode) (*os.File, error)
 	if name == "" {
 		return nil, &os.PathError{Op: "open", Path: name, Err: syscall.ENOENT}
 	}
-	r, errf := syscallOpenFileSequential(name, flag, 0)
+	// Use FILE_FLAG_SEQUENTIAL_SCAN rather than FILE_ATTRIBUTE_NORMAL as implemented in golang.
+	//https://msdn.microsoft.com/en-us/library/windows/desktop/aa363858(v=vs.85).aspx
+	const fileFlagSequentialScan = 0x08000000 // FILE_FLAG_SEQUENTIAL_SCAN
+	r, errf := syscallOpenFileWithFlags(name, flag, fileFlagSequentialScan)
 	if errf == nil {
 		return r, nil
 	}
 	return nil, &os.PathError{Op: "open", Path: name, Err: errf}
 }
 
-func syscallOpenFileSequential(name string, flag int, _ os.FileMode) (file *os.File, err error) {
-	r, e := syscallOpenSequential(name, flag|syscall.O_CLOEXEC, 0)
+// WriteFileTemporary writes data to a file named by filename.
+// If the file does not exist, WriteFile creates it with permissions perm;
+// otherwise WriteFile truncates it before writing.
+func WriteFileTemporary(filename string, data []byte, perm os.FileMode) error {
+	f, err := OpenFileTemporary(filename, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, perm)
+	if err != nil {
+		return err
+	}
+	n, err := f.Write(data)
+	if err == nil && n < len(data) {
+		err = io.ErrShortWrite
+	}
+	if err1 := f.Close(); err == nil {
+		err = err1
+	}
+	return err
+}
+
+// OpenFileTemporary creates the named file with mode 0666 (before umask), truncating
+// it if it already exists. If successful, methods on the returned
+// File can be used for I/O; the associated file descriptor has mode
+// O_RDWR.
+// If there is an error, it will be of type *PathError.
+func OpenFileTemporary(name string, flag int, _ os.FileMode) (*os.File, error) {
+	if name == "" {
+		return nil, &os.PathError{Op: "open", Path: name, Err: syscall.ENOENT}
+	}
+	// Use FILE_ATTRIBUTE_TEMPORARY rather than FILE_ATTRIBUTE_NORMAL as implemented in golang.
+	//https://msdn.microsoft.com/en-us/library/windows/desktop/aa363858(v=vs.85).aspx
+	const fileAttributeTemporary = 0x00000100 // FILE_ATTRIBUTE_TEMPORARY
```

Comment:
Hi @jhowardmsft,

Thanks for voicing your concerns.  The design of secrets in docker was to provide a solution
for the secure storage and distribution of sensitive material, such as API keys, passwords, 
TLS keys, etc. to applications running in containers, but tightly integrated into docker
swarm itself.

The word "secrets" for sensitive data that needs to be stored and transported securely,
but available to an application in plaintext, is an industry-standard term
(see [kubernetes secrets](https://kubernetes.io/docs/user-guide/secrets/),
[dc/os secrets](https://docs.mesosphere.com/1.8/administration/secrets/),
and [vault](https://www.vaultproject.io/intro/use-cases.html) for a couple examples)
and using shared terminology helps users understand what to expect from such a feature.

We wanted the interface to access secrets to be flexible, but in such a way
that the data wouldn't be inadverdently leaked.

The current common practice for injecting configuration values, environment
variables, can be inadvertently leaked in a variety of ways.  They are often
considered somewhat "public" by diagnostic tools; for example, the `-e` flag to
`ps` indiscriminately logs _all_ environment variables, including any
containing credentials.  This is one reason we chose files as the interface for
applications to request secrets: diagnostic tools and system administrators
routinely treat the contents of the filesystem as more sensitive, and will not
share it as a routine debugging step.

However, in addition to being secure, the interface to retrieve secrets must
also be relatively convenient for applications, or bad practices will arise for
applications to subvert the security of the secrets management layer to provide
a more convenient interface for themselves.  Most linux applications already
expect credentials to be supplied in plaintext, and often in the form of files
(e.g. a password or API key embedded in a config file, or a TLS key in a .pem
file) and would require changes in order to obtain secrets in another manner.

We are explicitly not dealing with malicious applications attempting to
exfiltrate the data through some other means.  Ultimately the _goal_ of the
secrets feature is to share a secret with an application.  If the application
itself is malicious, there's not much that can be done in many cases.

If applications already have logic to retrieve their secrets some other way, such as from an
API over Vault, we do not require a migration over to Docker secrets. In fact, the Docker
secrets feature can be used to securely bootstrap the use of Vault by distributing and managing
any Vault credentials. Users could also run a service with access to required secrets but
that provides an API interface for other containers obtain secrets, which would have the same
application security guarantees as Vault or any other current secrets distribution mechanism
that is not hardware-backed (for instance via a TPM or cloud HSM).

The secrets feature offers an advantage over directly writing files to the image itself
is because:

1. secrets should be specified at runtime, and not baked into an image, which is more general
2. since secrets exist only in a container mount, they cannot be committed into an image from a running
	container
3. the secrets would not be available in a stopped container (if the service were rescheduled
	to another worker, for instance)

Our trust model is that you trust the managers in which your secrets are stored, you trust
the workers to pass along the secrets that you require to run you services (but they only
have access to secrets that they need at any given time, and not all secrets), and you also
explicitly designate trust in your application by specifying exactly which secrets it is
allowed to have plaintext access to.

Compromise of the processes within a running container means compromising only the secrets
it has access to, compromising a running worker means compromising only the secrets that
all of the services it is supposed to be running has access to (since workers cannot schedule
or request extra secrets), and compromising a manager compromises all secrets.

Compromising the disk of a stopped worker, a stopped manager, or backups of the worker or
manager would ideally net you nothing, which is why we encrypt the swarm store at rest on
the managers, and never write the secrets to disk on the worker in linux.  The user should
really be using FDE and not provide any extraneous filesystem access to the daemon host,
but we aimed to provide defense in depth in case of an invalid configuration.

This part of the threat model is also why making the secrets feature expose its
own in-memory filesystem is important.  If we only present users with an API,
they may write utilities to integrate with existing software that requires
secrets to be supplied in disk files, and in doing so they might end up using a
less secure storage backend to hold those temporary files, such as a volume
(which may be shared and replicated with other hosts by a volume driver).

You are absolutely correct in that a system in which the application had zero access
to sensitive data, and hence requires very little trust, and would be ideal. We would love
to provide more support for such features in the future.  However, in reality, very few
applications (at least in linux) can currently use such a system, and we are attempting
to solve the more common use case.  Even Vault's
[dynamic secrets](https://www.vaultproject.io/intro/getting-started/dynamic-secrets.html)
feature (which we would love to be able to implement at least in part) presents the
generated secrets to the application in plaintext.  Our goal is to improve the state of
secrets distribution incrementally, which we feel is more valuable than waiting on an
ideal and can only work with newly written applications designed to work with such a
feature.

You are also correct that if swap is enabled, creating excessive memory
pressure can cause memory to be dumped to disk in plaintext if swap is enabled
and encrypted-swap is disabled.  This is an issue we would like to solve
without causing undue strain on a system's memory (for instance, by `mlockall`
the entire daemon process), but again we are attempting to solve the issue
incrementally, rather than waiting for the ideal.  Even in this case, though,
there are valid threats that a tmpfs can defend against - for example,
compromise of a stale backup might contain current secrets, but swap files are
rarely backed up and shared as widely as filesystem snapshots.  If a RAMDisk is
not available on windows, then perhaps we can further discuss other possible
solutions?

Thanks for reading.